---
title: "cS2G"
author: "<h4>Author: <i>Brian M. Schilder</i></h4>" 
date: "<h4>Most recent update: <i>`r format( Sys.Date(), '%b-%d-%Y')`</i></h4>"
output:
  BiocStyle::html_document:
vignette: >
    %\VignetteIndexEntry{cS2G} 
    %\usepackage[utf8]{inputenc}
    %\VignetteEngine{knitr::rmarkdown} 
---

```{r, include = FALSE}
knitr::opts_chunk$set(echo = TRUE, fig.width = 7, fig.height = 5)
```

```{r setup, include=TRUE, message=FALSE} 
library(phenomix)  
set.seed(2020)
```

# cS2G-mapped UK Biobank

## Import data

47 UKB traits that have been fine-mapped and mapped to genes using the
combined SNPs to Genes (cS2G) model.

```{r, eval=FALSE}
dataset <- "finemapping_cS2G_UKBB"
ukb_cs2g <- phenomix::get_cs2g(dataset = dataset,
                               value_var = "cS2G", 
                               min_value = .75, 
                               as_matrix = TRUE)
ukb_pip <- phenomix::get_cs2g(dataset = dataset,
                              value_var = "PIP",
                              as_matrix = TRUE)
```

## Decompose

### UMAP

```{r, fig.height=7, fig.width=8, eval=FALSE}
umap <- phenomix::run_umap(mat = ukb_cs2g$data)

gg_umap <- phenomix::plot_reduction(obj = umap, 
                                    obs = ukb_cs2g$obs, 
                                    size_var = "N_GENE", 
                                    color_var = "GROUP",
                                    label_var = "TRAIT") 
```


### MOFA2

Integrate the cS2G and PIP scores with MOFA2.

## Run MOFA2

*Note*: We've set `maxiter=10` for fast demonstration purposes only. The recommended default is `1000`.

```{r, eval=FALSE}
res <- phenomix::run_mofa2(obj = list(cS2G=ukb_cs2g$data, 
                                        PIP=ukb_pip$data), 
                           obs = ukb_cs2g$obs,
                           obs_idcol = "DISEASE.TRAIT", 
                           ### Set  low maxiter for demo 
                           maxiter = 10)
```

### Evaluate model 

```{r, eval=FALSE}
model <- res$model
MOFA2::plot_data_overview(model)

head(model@cache$variance_explained$r2_total[[1]]) # group 1
head(model@cache$variance_explained$r2_per_factor[[1]]) # group 1

MOFA2::plot_variance_explained(model, x="view", y="factor")

MOFA2::plot_factor(model, 
  factor = seq(3),
  color_by = "GROUP"
)

MOFA2::plot_factors(model, 
  factors = seq(3),
  color_by = "GROUP"
)

MOFA2::plot_weights(model,
  view = "PIP",
  factor = 1,
  nfeatures = 10,     # Number of features to highlight
  scale = TRUE,          # Scale weights from -1 to 1
  abs = FALSE             # Take the absolute value?
)

MOFA2::plot_top_weights(model,
  view = "PIP",
  factor = 1,
  nfeatures = 10
)

MOFA2::plot_data_heatmap(model,
  view = "PIP",         # view of interest
  factor = 1,             # factor of interest
  features = 20,          # number of features to plot (they are selected by weight)
  
  # extra arguments that are passed to the `pheatmap` function
  cluster_rows = TRUE, cluster_cols = FALSE,
  show_rownames = TRUE, show_colnames = FALSE
)
```

### Plot MOFA2 UMAP

Run UMAP on the MOFA factors and visualize.

```{r, eval=FALSE} 
MOFA2::plot_dimred(object = model,
                   method = "UMAP", 
                   color_by = "GROUP")
```


# cS2G-mapped GWAS Catalog

The larger GWAS Catalog dataset is also provided by `phenomix`. 

```{r, eval=FALSE} 
gwascatalog <- phenomix::get_cs2g(dataset = "gwas_catalog_cS2G", 
                                  value_var = "cS2G",
                                  as_matrix = TRUE)
```

## Run MOFA2

*Note*: We've set `maxiter=10` for fast demonstration purposes only. The recommended default is `1000`.

```{r, eval=FALSE}
res2 <- phenomix::run_mofa2(obj = list(GWAScatalog=gwascatalog$data), 
                            obs = gwascatalog$obs,
                            obs_idcol = "ID", 
                            ### Set  low maxiter for demo
                            maxiter = 10) 
```

### Plot MOFA2 UMAP

Run UMAP on the MOFA factors and visualize.

```{r, eval=FALSE}
MOFA2::plot_dimred(object = res2$model,
                   method = "UMAP", # TSNE 
                   color_by = "N_GENE")
```

# Session Info 

<details> 

```{r Session Info}
utils::sessionInfo()
```

</details>  

