<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>celltype_enrichment â€¢ phenomix</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="celltype_enrichment">
<meta property="og:description" content="phenomix">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">phenomix</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/celltype_enrichment.html">celltype_enrichment</a>
    </li>
    <li>
      <a href="../articles/hexSticker.html">hexSticker</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/neurogenomics/phenomix/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="celltype_enrichment_files/header-attrs-2.9/header-attrs.js"></script><script src="celltype_enrichment_files/htmlwidgets-1.5.3/htmlwidgets.js"></script><script src="celltype_enrichment_files/jquery-3.5.1/jquery.min.js"></script><link href="celltype_enrichment_files/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet">
<script src="celltype_enrichment_files/datatables-binding-0.18/datatables.js"></script><link href="celltype_enrichment_files/dt-core-1.10.20/css/jquery.dataTables.min.css" rel="stylesheet">
<link href="celltype_enrichment_files/dt-core-1.10.20/css/jquery.dataTables.extra.css" rel="stylesheet">
<script src="celltype_enrichment_files/dt-core-1.10.20/js/jquery.dataTables.min.js"></script><script src="celltype_enrichment_files/jszip-1.10.20/jszip.min.js"></script><script src="celltype_enrichment_files/pdfmake-1.10.20/pdfmake.js"></script><script src="celltype_enrichment_files/pdfmake-1.10.20/vfs_fonts.js"></script><link href="celltype_enrichment_files/dt-ext-buttons-1.10.20/css/buttons.dataTables.min.css" rel="stylesheet">
<script src="celltype_enrichment_files/dt-ext-buttons-1.10.20/js/dataTables.buttons.min.js"></script><script src="celltype_enrichment_files/dt-ext-buttons-1.10.20/js/buttons.flash.min.js"></script><script src="celltype_enrichment_files/dt-ext-buttons-1.10.20/js/buttons.html5.min.js"></script><script src="celltype_enrichment_files/dt-ext-buttons-1.10.20/js/buttons.colVis.min.js"></script><script src="celltype_enrichment_files/dt-ext-buttons-1.10.20/js/buttons.print.min.js"></script><link href="celltype_enrichment_files/crosstalk-1.1.1/css/crosstalk.css" rel="stylesheet">
<script src="celltype_enrichment_files/crosstalk-1.1.1/js/crosstalk.min.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>celltype_enrichment</h1>
                        <h4 class="author"><h4>
Author: <i>Brian M. Schilder</i>
</h4></h4>
            
            <h4 class="date"><h4>
Most recent update: <i>Jul-26-2021</i>
</h4></h4>
      
      <small class="dont-index">Source: <a href="https://github.com/neurogenomics/phenomix/blob/master/vignettes/celltype_enrichment.Rmd"><code>vignettes/celltype_enrichment.Rmd</code></a></small>
      <div class="hidden name"><code>celltype_enrichment.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/neurogenomics/phenomix">phenomix</a></span><span class="op">)</span>

<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span></code></pre></div>
<div id="import-data" class="section level1">
<h1 class="hasAnchor">
<a href="#import-data" class="anchor"></a>Import data</h1>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">### DeGAs loadings </span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"DEGAS_contributionGene"</span><span class="op">)</span>
<span class="va">xmat</span> <span class="op">&lt;-</span> <span class="va">DEGAS_contributionGene</span><span class="op">@</span><span class="va">feature.loadings</span>
<span class="va">xmat</span> <span class="op">&lt;-</span> <span class="va">xmat</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span> <span class="co"># Let's use just 10 components as an example</span>

<span class="co">### Celltype Dataset</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"ctd_BlueLake2018_FrontalCortexOnly"</span><span class="op">)</span>
<span class="va">ymat</span> <span class="op">&lt;-</span> <span class="va">ctd_BlueLake2018_FrontalCortexOnly</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">specificity</span></code></pre></div>
</div>
<div id="enrichment-tests" class="section level1">
<h1 class="hasAnchor">
<a href="#enrichment-tests" class="anchor"></a>Enrichment tests</h1>
<div id="linear-regression" class="section level2">
<h2 class="hasAnchor">
<a href="#linear-regression" class="anchor"></a>Linear regression</h2>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">res_lm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/iterate_lm.html">iterate_lm</a></span><span class="op">(</span>xmat <span class="op">=</span> <span class="va">xmat</span>, 
                     ymat <span class="op">=</span> <span class="va">ymat</span><span class="op">)</span> </code></pre></div>
<pre><code>## + 10 core(s) assigned as workers (2 reserved).</code></pre>
<pre><code>## 11410 intersecting genes between xmat and ymat</code></pre>
<pre><code>## Running 80 tests: 10 traits x 8 celltypes.</code></pre>
<pre><code>## 
## 15 significant results @ BH&lt;0.05</code></pre>
<div id="results" class="section level3">
<h3 class="hasAnchor">
<a href="#results" class="anchor"></a>Results</h3>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/create_DT.html">create_DT</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/subset.html">subset</a></span><span class="op">(</span><span class="va">res_lm</span>, <span class="va">qvalue</span><span class="op">&lt;</span><span class="fl">.05</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<div id="htmlwidget-43b65bb112a29df180e5" style="width:100%;height:auto;" class="datatables html-widget"></div>
<script type="application/json" data-for="htmlwidget-43b65bb112a29df180e5">{"x":{"filter":"none","extensions":["Buttons"],"caption":"<caption><\/caption>","data":[["1","2","3","4","5","6","7","8","9","10","11","12","13","14","15"],["contributionGene_3","contributionGene_3","contributionGene_3","contributionGene_4","contributionGene_5","contributionGene_5","contributionGene_5","contributionGene_5","contributionGene_7","contributionGene_7","contributionGene_9","contributionGene_9","contributionGene_9","contributionGene_10","contributionGene_10"],["human.BlueLake2018_FrontalCortexOnly.Ex","human.BlueLake2018_FrontalCortexOnly.In","human.BlueLake2018_FrontalCortexOnly.OPC","human.BlueLake2018_FrontalCortexOnly.OPC","human.BlueLake2018_FrontalCortexOnly.End","human.BlueLake2018_FrontalCortexOnly.Ex","human.BlueLake2018_FrontalCortexOnly.In","human.BlueLake2018_FrontalCortexOnly.OPC","human.BlueLake2018_FrontalCortexOnly.Ex","human.BlueLake2018_FrontalCortexOnly.In","human.BlueLake2018_FrontalCortexOnly.Ex","human.BlueLake2018_FrontalCortexOnly.In","human.BlueLake2018_FrontalCortexOnly.OPC","human.BlueLake2018_FrontalCortexOnly.Ex","human.BlueLake2018_FrontalCortexOnly.Mic"],[-5.53834835728596e-05,-4.83827132199372e-05,7.57559014740826e-05,0.000271213148005512,2.80316214039303e-05,-5.59648737353438e-05,-4.40099007429097e-05,3.5309262308861e-05,-5.20733044784683e-05,-3.77056280544981e-05,-3.75014258976146e-05,-3.43135933522744e-05,4.74074871886127e-05,-7.98225468702104e-05,6.84780595241885e-05],[1.7552117390565e-05,1.47014462428443e-05,1.7946415879829e-05,6.72132073465497e-05,8.37635239093168e-06,1.15205537916821e-05,9.65103288865283e-06,1.17909593046355e-05,1.47676162361391e-05,1.23713580110451e-05,1.27581187816395e-05,1.06856765923117e-05,1.30466059580525e-05,2.22326457054316e-05,1.7144782478253e-05],[-3.15537335698486,-3.29101725236636,4.22122734596989,4.03511688717879,3.34651887786832,-4.85782842971929,-4.56012338271629,2.99460471337387,-3.52618213026387,-3.04781641763456,-2.93941658166593,-3.21117648057614,3.63370269179872,-3.5903305404048,3.99410488940577],[0.00160709427610667,0.00100128852243076,2.44857817194318e-05,5.49311687031579e-05,0.000820974381044945,1.20257626728643e-06,5.16559471691961e-06,0.00275390300452948,0.000423251646301408,0.00231038163179702,0.00329490026875663,0.00132559214651345,0.000280609888673623,0.000331645592525201,6.53466581912906e-05],[0.0107139618407111,0.00801030817944608,0.000652954179184847,0.00104554653106065,0.00729755005373284,9.62061013829145e-05,0.000206623788676784,0.0157365885973113,0.00423251646301408,0.0142177331187509,0.0175728014333687,0.00964067015646144,0.00374146518231498,0.00379023534314516,0.00104554653106065]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>trait<\/th>\n      <th>term<\/th>\n      <th>estimate<\/th>\n      <th>std.error<\/th>\n      <th>statistic<\/th>\n      <th>pvalue<\/th>\n      <th>qvalue<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"dom":"Bfrtip","buttons":["copy","csv","excel","pdf","print"],"scrollY":400,"scrollX":true,"scrollCollapse":true,"paging":false,"columnDefs":[{"className":"dt-center","targets":"_all"},{"orderable":false,"targets":0}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":[],"jsHooks":[]}</script>
</div>
<div id="plot" class="section level3">
<h3 class="hasAnchor">
<a href="#plot" class="anchor"></a>Plot</h3>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Only plot the top 20 significant results</span>
<span class="va">data_lm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html">subset</a></span><span class="op">(</span><span class="va">res_lm</span>, <span class="va">qvalue</span><span class="op">&lt;</span><span class="fl">.05</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html">slice_min</a></span><span class="op">(</span><span class="va">qvalue</span>, n <span class="op">=</span> <span class="fl">20</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span><span class="op">(</span><span class="va">term</span>, <span class="va">qvalue</span><span class="op">)</span>

<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">data_lm</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">trait</span>, y<span class="op">=</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">qvalue</span><span class="op">)</span>, fill<span class="op">=</span><span class="va">term</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html">geom_bar</a></span><span class="op">(</span>stat <span class="op">=</span> <span class="st">"identity"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_grid.html">facet_grid</a></span><span class="op">(</span>facets <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span><span class="op">(</span><span class="st">"_"</span>,<span class="st">"\n"</span>,<span class="va">term</span><span class="op">)</span> <span class="op">~</span><span class="va">.</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span><span class="op">(</span>x<span class="op">=</span><span class="st">"celltype"</span>, title<span class="op">=</span><span class="st">"Linear regession results"</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">45</span>, hjust<span class="op">=</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span> </code></pre></div>
<p><img src="celltype_enrichment_files/figure-html/unnamed-chunk-5-1.png" width="672"></p>
</div>
</div>
<div id="gene-set-enrichment-analysis" class="section level2">
<h2 class="hasAnchor">
<a href="#gene-set-enrichment-analysis" class="anchor"></a>Gene set enrichment analysis</h2>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">res_gsea</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/iterate_gsea.html">iterate_gsea</a></span><span class="op">(</span>xmat <span class="op">=</span> <span class="va">xmat</span>, 
                         ymat <span class="op">=</span> <span class="va">ymat</span><span class="op">)</span> </code></pre></div>
<pre><code>## 11410 intersecting genes between GWAS and CTD matrices.</code></pre>
<pre><code>## Running 80 tests: 10 traits x 8 celltypes.</code></pre>
<pre><code>## 
## 0 significant results @ BH&lt;0.05</code></pre>
<div id="results-1" class="section level3">
<h3 class="hasAnchor">
<a href="#results-1" class="anchor"></a>Results</h3>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/create_DT.html">create_DT</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/subset.html">subset</a></span><span class="op">(</span><span class="va">res_gsea</span>, <span class="va">qvalue</span><span class="op">&lt;</span><span class="fl">.05</span><span class="op">)</span><span class="op">)</span></code></pre></div>
</div>
<div id="plot-1" class="section level3">
<h3 class="hasAnchor">
<a href="#plot-1" class="anchor"></a>Plot</h3>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Only plot the top 20 significant results</span>
<span class="va">data_gsea</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html">subset</a></span><span class="op">(</span><span class="va">res_gsea</span>, <span class="va">qvalue</span><span class="op">&lt;</span><span class="fl">.05</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html">slice_min</a></span><span class="op">(</span><span class="va">qvalue</span>, n <span class="op">=</span> <span class="fl">20</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span><span class="op">(</span><span class="va">term</span>, <span class="va">qvalue</span><span class="op">)</span>

<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">data_gsea</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">trait</span>, y<span class="op">=</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">qvalue</span><span class="op">)</span>, fill<span class="op">=</span><span class="va">term</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html">geom_bar</a></span><span class="op">(</span>stat <span class="op">=</span> <span class="st">"identity"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_grid.html">facet_grid</a></span><span class="op">(</span>facets <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span><span class="op">(</span><span class="st">"_"</span>,<span class="st">"\n"</span>,<span class="va">term</span><span class="op">)</span> <span class="op">~</span><span class="va">.</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span><span class="op">(</span>x<span class="op">=</span><span class="st">"celltype"</span>, title<span class="op">=</span><span class="st">"GSEA results"</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">45</span>, hjust<span class="op">=</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span> </code></pre></div>
</div>
</div>
</div>
<div id="session-info" class="section level1">
<h1 class="hasAnchor">
<a href="#session-info" class="anchor"></a>Session Info</h1>
<details><div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">utils</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code>## R version 4.1.0 (2021-05-18)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 20.04.2 LTS
## 
## Matrix products: default
## BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.8.so
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=C             
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
## [1] ggplot2_3.3.5  dplyr_1.0.7    phenomix_0.1.0
## 
## loaded via a namespace (and not attached):
##   [1] colorspace_2.0-2              ellipsis_0.3.2               
##   [3] rprojroot_2.0.2               XVector_0.32.0               
##   [5] GenomicRanges_1.44.0          ggdendro_0.1.22              
##   [7] fs_1.5.0                      farver_2.1.0                 
##   [9] listenv_0.8.0                 DT_0.18                      
##  [11] bit64_4.0.5                   interactiveDisplayBase_1.30.0
##  [13] AnnotationDbi_1.54.1          fansi_0.5.0                  
##  [15] xml2_1.3.2                    codetools_0.2-18             
##  [17] cachem_1.0.5                  knitr_1.33                   
##  [19] jsonlite_1.7.2                RNOmni_1.0.0                 
##  [21] dbplyr_2.1.1                  png_0.1-7                    
##  [23] shiny_1.6.0                   EWCE_1.0.1                   
##  [25] BiocManager_1.30.16           compiler_4.1.0               
##  [27] httr_1.4.2                    assertthat_0.2.1             
##  [29] Matrix_1.3-4                  fastmap_1.1.0                
##  [31] limma_3.48.1                  later_1.2.0                  
##  [33] htmltools_0.5.1.1             prettyunits_1.1.1            
##  [35] tools_4.1.0                   gtable_0.3.0                 
##  [37] glue_1.4.2                    GenomeInfoDbData_1.2.6       
##  [39] reshape2_1.4.4                rappdirs_0.3.3               
##  [41] Rcpp_1.0.7                    Biobase_2.52.0               
##  [43] jquerylib_0.1.4               pkgdown_1.6.1                
##  [45] vctrs_0.3.8                   Biostrings_2.60.1            
##  [47] ExperimentHub_2.0.0           crosstalk_1.1.1              
##  [49] xfun_0.24                     stringr_1.4.0                
##  [51] globals_0.14.0                mime_0.11                    
##  [53] lifecycle_1.0.0               gtools_3.9.2                 
##  [55] XML_3.99-0.6                  future_1.21.0                
##  [57] AnnotationHub_3.0.1           zlibbioc_1.38.0              
##  [59] MASS_7.3-54                   scales_1.1.1                 
##  [61] ragg_1.1.3                    hms_1.1.0                    
##  [63] promises_1.2.0.1              MatrixGenerics_1.4.0         
##  [65] parallel_4.1.0                SummarizedExperiment_1.22.0  
##  [67] RColorBrewer_1.1-2            yaml_2.2.1                   
##  [69] curl_4.3.2                    memoise_2.0.0                
##  [71] gridExtra_2.3                 sass_0.4.0                   
##  [73] biomaRt_2.48.2                stringi_1.7.3                
##  [75] RSQLite_2.2.7                 highr_0.9                    
##  [77] BiocVersion_3.13.1            S4Vectors_0.30.0             
##  [79] desc_1.3.0                    GeneOverlap_1.28.0           
##  [81] caTools_1.18.2                BiocGenerics_0.38.0          
##  [83] filelock_1.0.2                BiocParallel_1.26.1          
##  [85] GenomeInfoDb_1.28.1           rlang_0.4.11                 
##  [87] pkgconfig_2.0.3               systemfonts_1.0.2            
##  [89] bitops_1.0-7                  matrixStats_0.59.0           
##  [91] evaluate_0.14                 lattice_0.20-44              
##  [93] purrr_0.3.4                   labeling_0.4.2               
##  [95] htmlwidgets_1.5.3             bit_4.0.4                    
##  [97] tidyselect_1.1.1              parallelly_1.27.0            
##  [99] plyr_1.8.6                    magrittr_2.0.1               
## [101] R6_2.5.0                      gplots_3.1.1                 
## [103] IRanges_2.26.0                generics_0.1.0               
## [105] ewceData_1.0.0                DelayedArray_0.18.0          
## [107] DBI_1.1.1                     pillar_1.6.1                 
## [109] withr_2.4.2                   KEGGREST_1.32.0              
## [111] RCurl_1.98-1.3                tibble_3.1.2                 
## [113] crayon_1.4.1                  KernSmooth_2.23-20           
## [115] utf8_1.2.1                    BiocFileCache_2.0.0          
## [117] rmarkdown_2.9                 progress_1.2.2               
## [119] grid_4.1.0                    data.table_1.14.0            
## [121] blob_1.2.1                    digest_0.6.27                
## [123] xtable_1.8-4                  HGNChelper_0.8.1             
## [125] httpuv_1.6.1                  textshaping_0.3.5            
## [127] stats4_4.1.0                  munsell_0.5.0                
## [129] bslib_0.2.5.1</code></pre>
</details>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Brian Schilder.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
